sway-lsp/src/server.rs

```diff
@@ -1,659 +1,120 @@
-        }
-
-        // Initalizing tracing library based on the user's config
-        let config = self.config.read();
-        if config.logging.level != LevelFilter::OFF {
-            let tracing_options = TracingSubscriberOptions {
-                log_level: Some(config.logging.level),
-                writer_mode: Some(TracingWriterMode::Stderr),
-                ..Default::default()
-            };
-            init_tracing_subscriber(tracing_options);
-        }
-
-        tracing::info!("Initializing the Sway Language Server");
-
-        Ok(InitializeResult {
-            server_info: None,
-            capabilities: capabilities(),
-            ..InitializeResult::default()
-        })
+impl LanguageServer for GlobalState {
+    async fn initialize(&self, params: InitializeParams) -> Result<InitializeResult> {
+        request::handle_initialize(self, params)
     }
 
-    // LSP-Server Lifecycle
     async fn initialized(&self, _: InitializedParams) {
         tracing::info!("Sway Language Server Initialized");
     }
 
-    async fn shutdown(&self) -> jsonrpc::Result<()> {
-        tracing::info!("Shutting Down the Sway Language Server");
-
-        let _ = self.sessions.iter().map(|item| {
-            let session = item.value();
-            session.shutdown();
-        });
-
-        Ok(())
+    async fn shutdown(&self) -> Result<()> {
+        self.shutdown_server()
     }
 
-    // Document Handlers
     async fn did_open(&self, params: DidOpenTextDocumentParams) {
-        match self.get_uri_and_session(&params.text_document.uri) {
-            Ok((uri, session)) => {
-                session.handle_open_file(&uri);
-                self.parse_project(uri, params.text_document.uri, session.clone())
-                    .await;
-            }
-            Err(err) => tracing::error!("{}", err.to_string()),
-        }
+        notification::handle_did_open_text_document(self, params).await;
     }
 
     async fn did_change(&self, params: DidChangeTextDocumentParams) {
-        let config = self.config.read().on_enter.clone();
-        match self.get_uri_and_session(&params.text_document.uri) {
-            Ok((uri, session)) => {
-                // handle on_enter capabilities if they are enabled
-                capabilities::on_enter(&config, &self.client, &session, &uri.clone(), &params)
-                    .await;
-
-                // update this file with the new changes and write to disk
-                match session.write_changes_to_file(&uri, params.content_changes) {
-                    Ok(_) => {
-                        self.parse_project(uri, params.text_document.uri.clone(), session)
-                            .await;
-                    }
-                    Err(err) => tracing::error!("{}", err.to_string()),
-                }
-            }
-            Err(err) => tracing::error!("{}", err.to_string()),
-        }
+        notification::handle_did_change_text_document(self, params).await;
```

Comment:
Just curious, what happens if we don't await these? Do they still execute fully or do they get interrupted?

Reply:
these are `async` functions that return a `Future` when called. In order for the `Future` to be executed we need to call `.await` on them.